Spring Boot user language support documentation using cloudstate paradox plugin (#12)
cloudstateio/springboot-support: docs/src/main/paradox/cdi.md

# Context Injection

As we saw in the Getting Started example, it is perfectly possible to inject any Bean available in Spring into a 
Cloudstate entity class.
It is also possible to use the annotations present in the javax.inject package 
([JSR330](https://jcp.org/en/jsr/detail?id=330)). 
The Cloudstate Springboot Support library already includes the necessary dependencies so you don't have to worry about it.

> You can annotate your entity classes with Spring @Component or @Service annotations but we have created a convenient 

Reviewer: Suggest that all the annotations are in code with backticks.

```suggestion
You can annotate your entity classes with Spring `@Component` or `@Service` annotations but we have created a convenient 
```

> annotation that we call @CloudstateEntityBean that can be used for that too.

Reviewer: ```suggestion
annotation that we call `@CloudstateEntityBean` that can be used for that too.
```

## JSR330

> The Cloudastate Springboot Support library supports JSR330 within the scope of the support provided by Spring itself to 

Reviewer: `Cloudastate` -> `Cloudstate` and suggest consistently use `Spring Boot` everywhere in the text (not `Springboot`):

```suggestion
The Cloudstate Spring Boot Support library supports JSR330 within the scope of the support provided by Spring itself to 
```

this specification.
Note that the Cloudstate Java Support library on which we depend allows you to bind Cloudstate and any other DI container 
you want. However, no specific module for any of these other containers has yet been made.

Feel free to contribute or suggest support for more runtimes.

## Injecting EntityId and Cloudstate Context Objects

> You can use the @EntityId annotation to access the managed entity's id.

Reviewer: ```suggestion
You can use the `@EntityId` annotation to access the managed entity's id.
```

It is also possible to have access to the EventSourcedEntityCreationContext created during the activation of the object 

> by Cloudstate. However for this you will need to annotate the Context property with the annotation @CloudstateContext 

Reviewer: ```suggestion
by Cloudstate. However for this you will need to annotate the Context property with the annotation `@CloudstateContext`
```

as in the example below:

```java
@EntityId
private String entityId;

@CloudstateContext
private EventSourcedContext context;
```

### Using properties instead constructors

It is currently not possible to inject Cloudstate's EntityId and Context properties via constructor. 

> This is because the life cycles of these priorities differ from the life cycle of objects managed directly by Spring.

Reviewer: `priorities` should be `properties`?
```suggestion
This is because the life cycles of these properties differ from the life cycle of objects managed directly by Spring.
```

This is obviously only an issue if you want to inject EntityId or EventSourcedEntityCreationContext.
Otherwise, if you want to inject only other Spring Context Beans, you can use injection via builders as normal.

The builders below would be perfectly acceptable:

```java
@EventSourcedEntity

> @CloudstateEntityBean

Reviewer 2: Just a thought: I asked myself if its needed to have a separate annotation like `@CloudstateEntityBean`here. I see that `@CloudstateEntityBean` is an alias for a `@Component` but I have to mention here twice that this really is a cloudstate entity of a kind.

^^ I know this is a Document PR, but it did not catch my eye before.

Author: No problem @marcellanz.
Thanks for the feedback.
You can use both approaches, do you think this is bad?

public final class ShoppingCartEntity {
    private final Map<String, Shoppingcart.LineItem> cart = new LinkedHashMap<>();

    @EntityId
    private String entityId;

    @CloudstateContext
    private EventSourcedContext context;

>     private final RuleService ruleService;

Reviewer 2: like the constructor injection here; different on other occasions (see my other comments)

Author: In this case `@CloudstateContext` is used to assist the discovery engine of the springboot-support and not Spring itself. This is necessary because this type of data cannot be managed through the life cycle of Spring objects. I think I mention this in the documentation in the Conventions and Restrictions section. It could not be declared without this annotation or otherwise as it would probably cause an NPE at some point in the execution when that property was referenced

    private final ShoppingCartTypeConverter typeConverter;
    
    @Autowired

>     public ShoppingCartEntity(RuleService ruleService, ShoppingCartTypeConverter typeConverter){

Reviewer 2: I usually don't `@Autowire` annotate constructors, as it is not necessary for a component, which is one here as `@CloudstateEntityBean` is an alias for a `@Component`. Nothing wrong, just a preference :)

Author: I am cautious when it comes to examples, for me to be explicit about what is expected to happen is better than having the magic that Spring does under the hood. This is because in this case I want to make it clear that you can use injection by constructor in a normal way. It is worth noting that this behavior mentioned is only valid for versions starting with 4.3, that is, older developers may not be familiar with this behavior. Of course, this may be unlikely, but you never know.

Reviewer 2: Agree.

        this.ruleService = ruleService;
        this.typeConverter = typeConverter;
    }
    
    //......
}
```

> @@@ note { title=Important } As you can see, the constructor injection constraint applies only to EntityId and CreationContext. 

Reviewer 2: does not render properly
<img width="407" alt="Screenshot 2020-06-14 at 23 16 34" src="https://user-images.githubusercontent.com/12616445/84604273-1857f880-ae95-11ea-8cd3-964d59f5b1c3.png">

try
```
@@@ note { title=Important }
As you can see, the constructor injection constraint applies only to EntityId and CreationContext. 
So, as in the example above, you can mix the approaches and get the best of both worlds together.
@@@
```

Reviewer: Yes, the format for _block directives_ in paradox is to have the `@@@ note {}` and ending `@@@` on their own lines, just like fenced code blocks.
